### ISSUE TITLE
token mismatch with label in sequence labeling using BERT

### ISSUE THREAD

**Issue by possible1402** [allenai/allennlp#3977]
Because BERT uses wordpiece tokenization and when do sequence labeling task, I find that token mismatch with label. I use a Customized DataReader to generate token and label manually. If the token is split in many wordpieces, I add label correspondingly. It worked, but In the process of prediction, I don't know how the combine all wordpieces to one token to produce only one label rather than produce labels for every wordpieces.
Below is my codes for Customized DataReader and Predictor.
```
class PosDatasetReader(DatasetReader):

    def __init__(self, tokenizer: Callable[[str], str] = lambda x: x,
                 token_indexers: Dict[str, TokenIndexer] = None,
                 max_seq_len: Optional[int] = config.max_seq_len) -> None:
        super().__init__(lazy=False)
        self.token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer}
        self.tokenizer = tokenizer
        self.max_seq_len = max_seq_len


    def text_to_instance(self, token: List[Token], id: str = None, labels: List[str] = None) -> Instance:
        sentence_field = TextField(token, self.token_indexers)
        fields = {"sentence": sentence_field}
        id_field = MetadataField(id)
        fields["id"] = id_field
        if labels:
            label_field = SequenceLabelField(labels=labels, sequence_field=sentence_field)
            fields["labels"] = label_field
        return Instance(fields)


    def _read(self, file_path: str) -> Iterator[Instance]:
        pattern = re.compile(r'\d+\n')
        all_words = []
        all_labels = []
        ids = []
        with open(file_path, encoding='utf-8') as f:
            words = []
            labels = []
            for line in f:
                if pattern.match(line):
                    ids.append(str(line.replace('\n', '')))
                if pattern.match(line) or line == "" or line == "\n":
                    if words:
                        all_words.append(words)
                        all_labels.append(labels)
                        words = []
                        labels = []
                else:
                    splits = line.split('\t')
                    words.append(splits[0])
                    if len(splits) > 1:
                        labels.append(''.join(splits[1]).replace("\n", ""))
                    else:
                        labels.append("O")
            if words:
                all_words.append(words)
                all_labels.append(labels)
                # pairs = line.strip().split()
                # sentence, tags = zip(*(pair.split('###') for pair in pairs))
        for all_words_, ids_, all_labels_ in zip(all_words, ids, all_labels):
            tokens = []
            label_ids = []
            for word, label in zip(all_words_, all_labels_):
                word_tokens = self.tokenizer(word)
                if len(word_tokens) == 0:
                    pass
                else:
                    mm = [Token(x) for x in word_tokens]
                    tokens.extend(mm)
                    label_ids.extend([label] + [label] * (len(mm) - 1))
            if len(tokens) > self.max_seq_len:
                tokens = tokens[:self.max_seq_len]
                label_ids = label_ids[:self.max_seq_len]

            yield self.text_to_instance(tokens, ids_,
                                        label_ids)

```

```
class Predictor:
    def __init__(self, model: Model, iterator: DataIterator,
                 cuda_device: int = -1) -> None:
        self.model = model
        self.iterator = iterator
        self.cuda_device = cuda_device

    def _extract_data(self, batch) -> np.ndarray:
        out_dict = self.model(**batch)
        return expit(tonp(out_dict["class_logits"]))##shape:[batch size,max seq len,num labels]

    def predict(self, ds: Iterable[Instance]) -> np.ndarray:
        pred_generator = self.iterator(ds, num_epochs=1, shuffle=False)
        self.model.eval()
        pred_generator_tqdm = tqdm(pred_generator,
                                   total=self.iterator.get_num_batches(ds))
        preds = []
        with torch.no_grad():
            for batch in pred_generator_tqdm:
                batch = nn_util.move_to_device(batch, self.cuda_device)
                preds.append(self._extract_data(batch))
        return np.concatenate(preds, axis=0)
```

By the way, huggingface transformers handles this by converting the corresponding label ids of wordpieces except the first one to pad_token_label_id, so when in the process of training and prediction, the model can ignore the pad_token_label_id. But in allennlp, because iterator indexes token and label id automatically, I don't know how to reproduce the method that huggingface transformers uses.

**Comment by matt-gardner** [comment]
This is precisely what our [`PretrainedTransformerMismatchedIndexer`](https://github.com/allenai/allennlp/blob/master/allennlp/data/token_indexers/pretrained_transformer_mismatched_indexer.py) is used for (along with the corresponding `PretrainedTransformerMismatchedEmbedder`).   If you need help using that, let us know, but it should be straightforward.  Then you don't have to do any messing around with the labels, because your model ends up with the same word-level tokenization as you had in your input, which correspond to the original labels in the data.

I wasn't aware of what huggingface transformers do with padding label ids, but it seems broken to me.  For example, if you want to use a CRF, you mathematically *can't* model transitions between labels correctly.  You can imagine distributing the labels across all of the wordpieces in a word, and then at least you still have correct label transitions, but inserting padded labels in between is just broken, unless they have something very complex on top to fix the problem (which would end up doing something very similar to what we do in our mismatched case, anyway).

**Comment by possible1402** [comment]
@matt-gardner Thanks for your quick reply!
I deside to use PretrainedTransformerMismatchedIndexer, but it seems that allennlp deletes `iterators`. What should I do without iterators? How to index_with vocab and set batch_size?

**Comment by matt-gardner** [comment]
See here: https://github.com/allenai/allennlp/releases/tag/v1.0-prerelease.

**Comment by possible1402** [comment]
@matt-gardner I tried my best to implement but failed. Sorry to bother you again~ 
My previous code is:
`vocab = Vocabulary.from_instances(train_ds + val_ds + test_ds)
iterator = BucketIterator(batch_size=config.batch_size,
                          sorting_keys=[("sentence", "num_tokens")],
                          )
iterator.index_with(vocab)`
 I use this method to index token and label but now iterator is deleted. I try to use `TextField.index` but it doesn't work. And I find that there is a function in `PretrainedTransformerMismatchedIndexer` named `tokens_to_indices`. I can only index token in one instance of the train dataset. 
`token_indexer.tokens_to_indices(train_ds.instances[0].fields['sentence'],vocab)`
I have a few questions.
1. How can I index token in all of the training dataset?
2. Can I just use `vocab = Vocabulary()` as vocabulary to index token and label? Because I used to use `vocab = Vocabulary.from_instances(train_ds + val_ds + test_ds)` to create namespece for labels or it will occur the error `KeyError: '@@UNKNOWN@@'`


below is a part of my codes:
`class PosDatasetReader(DatasetReader):


    def __init__(self,
                 token_indexers: Dict[str, TokenIndexer] = None,
                 max_seq_len: Optional[int] = config.max_seq_len,
                 ) -> None:
        super().__init__(lazy=False)
        self.token_indexers = token_indexers or {"tokens": SingleIdTokenIndexer}
        self.max_seq_len = max_seq_len
        self.vocab=vocab


     def text_to_instance(self, token: List[Token], id: str = None, labels: List[str] = None) -> Instance:
        sentence_field = TextField(token, self.token_indexers)
        fields = {"sentence": sentence_field}
        id_field = MetadataField(id)
        fields["id"] = id_field
        if labels:
            label_field = SequenceLabelField(labels=labels, sequence_field=sentence_field)
            fields["labels"] = label_field

        return Instance(fields)

    def _read(self, file_path: str) -> Iterator[Instance]:
        pattern = re.compile(r'\d+\n')
        all_words = []
        all_labels = []
        ids = []
        with open(file_path, encoding='utf-8') as f:
            words = []
            labels = []
            for line in f:
                if pattern.match(line):
                    ids.append(str(line.replace('\n', '')))
                if pattern.match(line) or line == "" or line == "\n":
                    if words:
                        all_words.append(words)
                        all_labels.append(labels)
                        words = []
                        labels = []
                else:
                    splits = line.split('\t')
                    words.append(splits[0])
                    if len(splits) > 1:
                        labels.append(''.join(splits[1]).replace("\n", ""))
                    else:
                        labels.append("O")
            if words:
                all_words.append(words)
                all_labels.append(labels)
                # pairs = line.strip().split()
                # sentence, tags = zip(*(pair.split('###') for pair in pairs))
        for all_words_, ids_, all_labels_ in zip(all_words, ids, all_labels):
            yield self.text_to_instance([Token(word) for word in all_words_], ids_, all_labels_)


vocab = Vocabulary()
token_indexer = PretrainedTransformerMismatchedIndexer(model_name='bert-base-chinese', max_length=config.max_seq_len)
DATA_ROOT = Path("data_for_test")
reader = PosDatasetReader(token_indexers={"tokens": token_indexer})
train_ds, val_ds, test_ds = (reader.read(DATA_ROOT / fname) for fname in ["train.tsv", "dev.tsv", "test.tsv"])
len(train_ds)
print(train_ds[0].fields['labels'].labels)
print(train_ds[0].fields['labels'].sequence_field)`


I'm really grateful for your quick response!

**Comment by matt-gardner** [comment]
You should take your old code, and just replace the `iterator = ...` line with something like this:

```python
vocab = Vocabulary.from_instances(train_ds + val_ds + test_ds)  # as you had it before
train_ds.index_with(vocab)
val_ds.index_with(vocab)
test_ds.index_with(vocab)

train_sampler = RandomSampler(train_ds)
...  # the rest of your data loader code
```

**Comment by matt-gardner** [comment]
This is basically our script for creating datasets, vocab, model, data loaders, and a trainer: https://github.com/allenai/allennlp/blob/15df437c202a678aa6e8c16a3567f14bdcef547b/allennlp/commands/train.py#L628-L689

If you're writing your own script, it should look something very much like that, just without the `Lazy` objects and their calls to `.construct()`.  We use those to separate object configuration from object construction, so that we can handle config files properly.  If you're writing your own script, you probably aren't using config files, and you don't need the `Lazy` stuff.

**Comment by possible1402** [comment]
@matt-gardner Words fail me when I want to express my gratitude to you!
It worked exactly perfect! Sorry for one more question. 
The code runs corretly on my own machine without CUDA and I use the dataset of 5 samples and the batch size is 4, but when I train the model on the machine with a 12G GPU ,the dataset size is 100000 samples and the batch size is 7, I get the following error:
`  File "C:\Users\dell\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-6641e059df1a>", line 1, in <module>
    print(trainer.train())
  File "C:\Users\dell\Anaconda3\lib\site-packages\allennlp-0.9.1_unreleased-py3.7.egg\allennlp\training\trainer.py", line 687, in train
    train_metrics = self._train_epoch(epoch)
  File "C:\Users\dell\Anaconda3\lib\site-packages\allennlp-0.9.1_unreleased-py3.7.egg\allennlp\training\trainer.py", line 465, in _train_epoch
    batch_outputs = self.batch_outputs(batch, for_training=True)
  File "C:\Users\dell\Anaconda3\lib\site-packages\allennlp-0.9.1_unreleased-py3.7.egg\allennlp\training\trainer.py", line 380, in batch_outputs
    output_dict = self._pytorch_model(**batch)
  File "C:\Users\dell\Anaconda3\lib\site-packages\torch\nn\modules\module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "E:/ding/BERT_bio_keyphrase_extraction_pytorch/allennlp_pretrained_mismatch.py", line 230, in forward
    embeddings = self.word_embeddings(sentence)
  File "C:\Users\dell\Anaconda3\lib\site-packages\torch\nn\modules\module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Users\dell\Anaconda3\lib\site-packages\allennlp-0.9.1_unreleased-py3.7.egg\allennlp\modules\text_field_embedders\basic_text_field_embedder.py", line 89, in forward
    token_vectors = embedder(**tensors, **forward_params_values)
  File "C:\Users\dell\Anaconda3\lib\site-packages\torch\nn\modules\module.py", line 547, in __call__
    result = self.forward(*input, **kwargs)
  File "C:\Users\dell\Anaconda3\lib\site-packages\allennlp-0.9.1_unreleased-py3.7.egg\allennlp\modules\token_embedders\pretrained_transformer_mismatched_embedder.py", line 81, in forward
    span_embeddings *= span_mask  # zero out paddings
RuntimeError: expected device cuda:0 and dtype Float but got device cuda:0 and dtype Bool `
This code throws this error:
`trainer = GradientDescentTrainer(
    model=model,
    optimizer=optimizer,
    data_loader=train_dataloader,
    validation_data_loader=val_dataloader,
    cuda_device=0 if USE_GPU else -1,
    num_epochs=config.epochs,
)

print(trainer.train())`

I keep printing `trainer.train()` and throws the error `RuntimeError: CUDA out of memory.`
So I believe it's the memory error. But I don't think it's because of the training data size and batch size. Because I use my old code to train this model on the same GPU with the same training data size and batch size.
The only code in the traning process I change is from 'Trainer' to 'GradientDescentTrainer'. While I don't think it's owing to that. 
So I'm wondering whether it's owing to `PretrainedTransformerMismatchedEmbedder`

**Comment by matt-gardner** [comment]
@ZhaofengWu, this latest error seems related to the switch to bool masks.  Is this a place where we need a `.float()` in the code?

**Comment by ZhaofengWu** [comment]
@possible1402 What's your PyTorch version? Is it in compliance with our version specification? In the latest version float and bool are compatible in multiplication.

**Comment by possible1402** [comment]
@ZhaofengWu @matt-gardner I changed to torch1.4.0 and that error disappeared. But when run 1528 steps, it throws the error `RuntimeError: CUDA out of memory. Tried to allocate 48.00 MiB (GPU 0; 11.00 GiB total capacity; 8.54 GiB already allocated; 47.71 MiB free; 8.66 GiB reserved in total by PyTorch)`.
I used to use batch size 7 and can run perfectly, but now I only use batch size 2. And why does the code can run 1528 steps rather than throwing the memory error at the first time? I set the max_seq_len of 512. But my dataset has lots of examples that exceeds that length.

I'm wondering the `max_length `augument in the `PretrainedTransformerMismatchedIndexer`.
It's said that ` If positive, split the document into segments of this many tokens (including special tokens) before feeding into the embedder. The embedder embeds these segments independently and concatenate the results to get the original document representation. `.

So the splited examples are concatenated to the batch? If so, I think its's unreasonable, it might cause the memory error.
